rsmtool/rsmexplain.py

```diff
@@ -0,0 +1,525 @@
+        is a tuple, sub-sample the range of examples using the two values
+        in the tuple. If this is ``None``, use all of the examples without
+        any sub-sampling.
+
+    Returns
+    -------
+    Dict[int, str]
+        Dictionary mapping the position of the selected examples to their IDs.
+        This is useful for figuring out which specific examples were selected.
+    numpy.ndarray
+        A 2D numpy array containing sampled feature rows.
+    """
+    # get a sparse matrix with the features that were actually used
+    features = learner.feat_selector.transform(
+        learner.feat_vectorizer.transform(
+            featureset.vectorizer.inverse_transform(featureset.features)
+        )
+    )
+
+    # sample exampeles from the featureset
+    selected_feature_map = select_features(featureset, range_size=feature_range)
+
+    # if the user specified a sample size or a range, use it; otherwise all
+    # features will be selected
+    if feature_range:
+        positions = list(selected_feature_map.keys())
+        features = features[positions, :]
+
+    # convert to a dense array if not already one
+    features = features.toarray() if not isinstance(features, np.ndarray) else features
+    return selected_feature_map, features
+
+
+def generate_explanation(
+    config_file_or_obj_or_dict, output_dir, overwrite_output=False, logger=None
+):
+    """
+    Generate a shap.Explanation object.
+
+    This function does all the heavy lifting. It loads the model, creates an explainer, and
+    generates an explanation object. It then calls generate_report() in order to generate a SHAP
+    report.
+
+    Parameters
+    ----------
+    config_file_or_obj_or_dict : str or pathlib.Path or dict or Configuration
+        Path to the experiment configuration file either a a string
+        or as a ``pathlib.Path`` object. Users can also pass a
+        ``Configuration`` object that is in memory or a Python dictionary
+        with keys corresponding to fields in the configuration file. Given a
+        configuration file, any relative paths in the configuration file
+        will be interpreted relative to the location of the file. Given a
+        ``Configuration`` object, relative paths will be interpreted
+        relative to the ``configdir`` attribute, that _must_ be set. Given
+        a dictionary, the reference path is set to the current directory.
+    output_dir : str
+        Path to the experiment output directory.
+    logger : Optional[logging object]
+        A logging object. If ``None`` is passed, get logger from ``__name__``.
+        Defaults to ``None``.
+
+    Raises
+    ------
+    AssertionError
+        If the size of the background sample is < 300, to avoid inaccurate explanations.
+    FileNotFoundError
+        If any file contained in ``config_file_or_obj_or_dict`` cannot be located.
+    ValueError
+        If range is not an integer of list of integer.
```

Comment:
what is `range`? is in the configuration?

Reply:
looks like it raises an error only if both `sample_range` and `sample_size` are specified. not if none of them.